File: mobile/ios/Sources/ChatGPTOAuth.swift
import AuthenticationServices
import CryptoKit
import Foundation
import UIKit

/// The existing desktop GPT OAuth contract, adapted to a mobile callback. It is
/// a public PKCE client: no client secret is ever accepted or persisted.
@MainActor
final class ChatGPTOAuth: NSObject, ObservableObject, ASWebAuthenticationPresentationContextProviding {
    @Published private(set) var accessToken: String
    @Published private(set) var accountID: String
    @Published private(set) var status = ""
    @Published private(set) var isSigningIn = false

    private let secure = SecureStore()
    private var session: ASWebAuthenticationSession?
    private let clientID = "app_EMoamEEZ73f0CkXaXp7hrann"
    private let callback = "herness://oauth/openai"

    override init() {
        accessToken = secure.read("oauth.openai.access") ?? ""
        accountID = secure.read("oauth.openai.account") ?? ""
        super.init()
    }

    func start() {
        let verifier = Self.randomToken()
        let state = Self.randomToken()
        secure.write(verifier, key: "oauth.openai.verifier")
        secure.write(state, key: "oauth.openai.state")
        guard var components = URLComponents(string: "https://auth.openai.com/oauth/authorize") else { return }
        components.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "redirect_uri", value: callback),
            URLQueryItem(name: "scope", value: "openid profile email offline_access"),
            URLQueryItem(name: "code_challenge", value: Self.challenge(verifier)),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "id_token_add_organizations", value: "true"),
            URLQueryItem(name: "codex_cli_simplified_flow", value: "true"),
            URLQueryItem(name: "originator", value: "dots_harness"),
        ]
        guard let url = components.url else { return }
        isSigningIn = true
        status = "Waiting for ChatGPT authorization…"
        let auth = ASWebAuthenticationSession(url: url, callbackURLScheme: "herness") { [weak self] callbackURL, error in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.session = nil
                if let callbackURL { _ = self.handleCallback(callbackURL) }
                else if let error, (error as NSError).code != ASWebAuthenticationSessionError.canceledLogin.rawValue { self.status = error.localizedDescription; self.isSigningIn = false }
                else { self.status = "ChatGPT sign-in cancelled."; self.isSigningIn = false }
            }
        }
        auth.presentationContextProvider = self
        session = auth
        guard auth.start() else { session = nil; isSigningIn = false; status = "Could not open ChatGPT sign-in."; return }
    }

    @discardableResult
    func handleCallback(_ url: URL) -> Bool {
        guard url.scheme?.lowercased() == "herness", url.host?.lowercased() == "oauth", url.path == "/openai" else { return false }
        let values = Dictionary(uniqueKeysWithValues: (URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []).compactMap { item in item.value.map { (item.name, $0) } })
        if let error = values["error"] { status = values["error_description"] ?? error; isSigningIn = false; return true }
        guard values["state"] == secure.read("oauth.openai.state"),
              let code = values["code"], let verifier = secure.read("oauth.openai.verifier") else {
            status = "ChatGPT OAuth state mismatch. Start sign-in again."
            isSigningIn = false
            return true
        }
        secure.delete("oauth.openai.state")
        secure.delete("oauth.openai.verifier")
        isSigningIn = true
        Task { await exchange(code: code, verifier: verifier) }
        return true
    }

    func refreshIfNeeded() async {
        let expiry = (Self.jwtClaims(accessToken)["exp"] as? NSNumber)?.doubleValue
        guard !accessToken.isEmpty, let expiry, expiry > Date().timeIntervalSince1970 + 60 else {
            if !accessToken.isEmpty && expiry == nil { return }
            guard let refreshToken = secure.read("oauth.openai.refresh"), !refreshToken.isEmpty else { return }
            await refresh(refreshToken)
            return
        }
    }

    func signOut() {
        secure.delete("oauth.openai.access")
        secure.delete("oauth.openai.refresh")
        secure.delete("oauth.openai.account")
        accessToken = ""; accountID = ""; status = "Signed out of ChatGPT on this device."
    }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.flatMap(\.windows).first(where: \.isKeyWindow) ?? UIWindow()
    }

    private func exchange(code: String, verifier: String) async {
        await requestToken(fields: [("grant_type", "authorization_code"), ("code", code), ("redirect_uri", callback), ("client_id", clientID), ("code_verifier", verifier)])
    }

    private func refresh(_ token: String) async {
        await requestToken(fields: [("grant_type", "refresh_token"), ("refresh_token", token), ("client_id", clientID)])
    }

    private func requestToken(fields: [(String, String)]) async {
        defer { isSigningIn = false }
        guard let url = URL(string: "https://auth.openai.com/oauth/token") else { return }
        var request = URLRequest(url: url, timeoutInterval: 30)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = Self.form(fields)
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
            guard (response as? HTTPURLResponse).map({ (200..<300).contains($0.statusCode) }) == true, let value = object["access_token"] as? String, !value.isEmpty else {
                status = object["error_description"] as? String ?? "ChatGPT token exchange failed."
                return
            }
            accessToken = value
            if let refresh = object["refresh_token"] as? String { secure.write(refresh, key: "oauth.openai.refresh") }
            let claims = Self.jwtClaims(object["id_token"] as? String ?? value)
            let auth = claims["https://api.openai.com/auth"] as? [String: Any]
            accountID = auth?["chatgpt_account_id"] as? String ?? claims["chatgpt_account_id"] as? String ?? object["account_id"] as? String ?? accountID
            secure.write(value, key: "oauth.openai.access")
            if !accountID.isEmpty { secure.write(accountID, key: "oauth.openai.account") }
            MobileStateStore().saveOAuth(provider: "gpt", accountReference: "keychain:oauth.openai.access")
            status = "Signed in to ChatGPT."
        } catch { status = error.localizedDescription }
    }

    private static func randomToken() -> String { base64URL(Data((0..<32).map { _ in UInt8.random(in: 0...255) })) }
    private static func challenge(_ value: String) -> String { base64URL(Data(SHA256.hash(data: Data(value.utf8)))) }
    private static func base64URL(_ data: Data) -> String { data.base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").trimmingCharacters(in: CharacterSet(charactersIn: "=")) }
    private static func jwtClaims(_ token: String) -> [String: Any] {
        let parts = token.split(separator: ".")
        guard parts.count > 1 else { return [:] }
        let raw = String(parts[1]).replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let padded = raw + String(repeating: "=", count: (4 - raw.count % 4) % 4)
        guard let data = Data(base64Encoded: padded, options: .ignoreUnknownCharacters) else { return [:] }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
    }
    private static func form(_ values: [(String, String)]) -> Data {
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-._~"))
        return Data(values.map { ($0.0.addingPercentEncoding(withAllowedCharacters: allowed) ?? $0.0) + "=" + ($0.1.addingPercentEncoding(withAllowedCharacters: allowed) ?? $0.1) }.joined(separator: "&").utf8)
    }
}

File: mobile/android/app/src/main/java/com/dots/herness/mobile/ChatGPTOAuth.kt
package com.dots.herness.mobile

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.security.MessageDigest
import java.util.Base64

/// Existing desktop GPT OAuth contract, adapted to a mobile PKCE callback.
class ChatGPTOAuth(context: Context) {
    private val appContext = context.applicationContext
    private val secure = SecureStore(appContext)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    var accessToken by mutableStateOf(secure.read("oauth.openai.access").orEmpty()); private set
    var accountId by mutableStateOf(secure.read("oauth.openai.account").orEmpty()); private set
    var status by mutableStateOf(""); private set
    var signingIn by mutableStateOf(false); private set

    fun start(activity: Activity?) {
        val verifier = randomToken()
        val state = randomToken()
        secure.write("oauth.openai.verifier", verifier)
        secure.write("oauth.openai.state", state)
        val uri = Uri.parse("https://auth.openai.com/oauth/authorize").buildUpon()
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("client_id", CLIENT_ID)
            .appendQueryParameter("redirect_uri", CALLBACK)
            .appendQueryParameter("scope", "openid profile email offline_access")
            .appendQueryParameter("code_challenge", challenge(verifier))
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("state", state)
            .appendQueryParameter("id_token_add_organizations", "true")
            .appendQueryParameter("codex_cli_simplified_flow", "true")
            .appendQueryParameter("originator", "dots_harness")
            .build()
        signingIn = true
        status = "Waiting for ChatGPT authorization…"
        val intent = Intent(Intent.ACTION_VIEW, uri)
        activity?.startActivity(intent) ?: run {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            appContext.startActivity(intent)
        }
    }

    fun handleCallback(uri: Uri): Boolean {
        if (!uri.scheme.equals("herness", true) || !uri.host.equals("oauth", true) || uri.path != "/openai") return false
        uri.getQueryParameter("error")?.let {
            status = uri.getQueryParameter("error_description") ?: it
            signingIn = false
            return true
        }
        val expected = secure.read("oauth.openai.state")
        if (expected.isNullOrBlank() || uri.getQueryParameter("state") != expected) {
            status = "ChatGPT OAuth state mismatch. Start sign-in again."
            signingIn = false
            return true
        }
        val code = uri.getQueryParameter("code").orEmpty()
        val verifier = secure.read("oauth.openai.verifier").orEmpty()
        if (code.isBlank() || verifier.isBlank()) {
            status = "ChatGPT did not return an authorization code."
            signingIn = false
            return true
        }
        secure.delete("oauth.openai.state")
        secure.delete("oauth.openai.verifier")
        signingIn = true
        scope.launch { requestToken(listOf("grant_type" to "authorization_code", "code" to code, "redirect_uri" to CALLBACK, "client_id" to CLIENT_ID, "code_verifier" to verifier)) }
        return true
    }

    suspend fun refreshIfNeeded() {
        val claims = jwtClaims(accessToken)
        val expiry = claims.optLong("exp", 0)
        if (accessToken.isNotBlank() && (expiry == 0L || expiry > System.currentTimeMillis() / 1000 + 60)) return
        val refresh = secure.read("oauth.openai.refresh").orEmpty()
        if (refresh.isNotBlank()) requestToken(listOf("grant_type" to "refresh_token", "refresh_token" to refresh, "client_id" to CLIENT_ID))
    }

    fun signOut() {
        secure.delete("oauth.openai.access")
        secure.delete("oauth.openai.refresh")
        secure.delete("oauth.openai.account")
        accessToken = ""; accountId = ""; status = "Signed out of ChatGPT on this device."
    }

    private suspend fun requestToken(fields: List<Pair<String, String>>) {
        try {
            val result = withContext(Dispatchers.IO) {
                val connection = (URL("https://auth.openai.com/oauth/token").openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    connectTimeout = 30_000
                    readTimeout = 30_000
                    doOutput = true
                    setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                    outputStream.use { it.write(fields.joinToString("&") { (key, value) -> URLEncoder.encode(key, "UTF-8") + "=" + URLEncoder.encode(value, "UTF-8") }.toByteArray()) }
                }
                val statusCode = connection.responseCode
                val bytes = (if (statusCode in 200..299) connection.inputStream else connection.errorStream)?.use { it.readBytes() } ?: ByteArray(0)
                statusCode to JSONObject(String(bytes))
            }
            val token = result.second.optString("access_token")
            if (result.first !in 200..299 || token.isBlank()) {
                status = result.second.optString("error_description").ifBlank { "ChatGPT token exchange failed." }
                return
            }
            accessToken = token
            result.second.optString("refresh_token").takeIf(String::isNotBlank)?.let { secure.write("oauth.openai.refresh", it) }
            val claims = jwtClaims(result.second.optString("id_token").ifBlank { token })
            accountId = claims.optJSONObject("https://api.openai.com/auth")?.optString("chatgpt_account_id").orEmpty()
                .ifBlank { claims.optString("chatgpt_account_id") }
                .ifBlank { result.second.optString("account_id") }
                .ifBlank { accountId }
            secure.write("oauth.openai.access", token)
            if (accountId.isNotBlank()) secure.write("oauth.openai.account", accountId)
            MobileStateStore(appContext).saveOAuth("gpt", "keystore:oauth.openai.access")
            status = "Signed in to ChatGPT."
        } catch (t: Throwable) {
            status = t.message ?: "ChatGPT token exchange failed."
        } finally {
            signingIn = false
        }
    }

    private fun randomToken(): String = Base64.getUrlEncoder().withoutPadding().encodeToString(ByteArray(32).also { java.security.SecureRandom().nextBytes(it) })
    private fun challenge(value: String): String = Base64.getUrlEncoder().withoutPadding().encodeToString(MessageDigest.getInstance("SHA-256").digest(value.toByteArray()))
    private fun jwtClaims(token: String): JSONObject {
        val segment = token.split(".").getOrNull(1) ?: return JSONObject()
        return runCatching { JSONObject(String(Base64.getUrlDecoder().decode(segment))) }.getOrElse { JSONObject() }
    }

    companion object {
        private const val CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"
        private const val CALLBACK = "herness://oauth/openai"
    }
}
